### PR TITLE
♻️(api) refactor indicators preprocessing

### PR DESCRIPTION
**Pull request by jmaupetit**
## Purpose

Data processing functions should not rely on a class instance attribute to apply processing on.

## Proposal

It's way more explicit to inject data to process as an argument and return processed data.

### CODE REVIEW

File: src/api/core/warren/indicators/base.py [hunk 1]
```diff
@@ -19,23 +19,19 @@ class PreprocessMixin(object):
...
+    def add_date_column(self, statements: pd.DataFrame) -> pd.DataFrame:
         """Add a 'date' column computed from statement's timestamp."""
         # Disable chained assignment warning to make the transformation inplace
         with pd.option_context("mode.chained_assignment", None):
             # Transform 'timestamp' column into a date with the format (YYYY-MM-DD)
-            self.statements["date"] = pd.to_datetime(
-                self.statements["timestamp"]
-            ).dt.date
+            statements["date"] = pd.to_datetime(statements["timestamp"]).dt.date
+        return statements.copy()
```
Comment (lebaudantoine) [inline src/api/core/warren/indicators/base.py:32]: Here's a quick question: should it occur before modifying the `statements`?

File: src/api/core/warren/indicators/base.py [hunk 2]
```diff
@@ -44,9 +40,7 @@ def add_actor_uid_column(self) -> None:
         without worrying about the 4 IFIs.
         """
         xapi_actor_identifier_columns = (
-            settings.XAPI_ACTOR_IDENTIFIER_PATHS.intersection(
-                set(self.statements.columns)
-            )
+            settings.XAPI_ACTOR_IDENTIFIER_PATHS.intersection(set(statements.columns))
```
Comment (lebaudantoine) [inline src/api/core/warren/indicators/base.py:43]:
```suggestion
            settings.XAPI_ACTOR_IDENTIFIER_PATHS.intersection(statements.columns)
```

File: src/api/core/warren/indicators/base.py [hunk 3]
```diff
@@ -61,13 +55,16 @@ def get_uid(row):
...
+    def preprocess_statements(self, statements: List[XAPI_STATEMENT]) -> pd.DataFrame:
         """Denormalize raw statements, and add utility columns."""
-        self.parse_raw_statements()
-        self.add_actor_uid_column()
-        self.add_date_column()
+        return compose(
+            self.add_date_column,
+            self.add_actor_uid_column,
+            self.normalize_statements,
+        )(statements)
```
Comment (lebaudantoine) [inline src/api/core/warren/indicators/base.py:67]: This lacks clarity. Without delving into the compose documentation, it appears that the steps are applied in reverse order.

File: src/api/core/warren/indicators/base.py [hunk 4]
```diff
@@ -78,16 +75,23 @@ class BaseIndicator(ABC):
...
-    async def fetch_statements(self) -> None:
-        """Execute the LRS query to obtain statements required for the indicator."""
+    async def fetch_statements(
+        self, lrs_query: LRSStatementsQuery = None
+    ) -> List[XAPI_STATEMENT]:
+        """Execute the LRS query to obtain statements required for this indicator.
 
-    @abstractmethod
-    def filter_statements(self) -> None:
-        """Filter the statements required for the indicator."""
```
Comment (lebaudantoine) [inline src/api/core/warren/indicators/base.py:90]: Nice 👍

File: src/api/core/warren/utils.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+"""Warren utils."""
+from functools import reduce
+
+
+def compose(*functions):
```
Comment (lebaudantoine) [inline src/api/core/warren/utils.py:5]: Nit:  could we have type hint?

File: src/api/core/warren/utils.py [hunk 6]
```diff
@@ -0,0 +1,10 @@
...
+    """Function composition.
+
+    Note that functions are applied in reverse order (last one first).
+    """
+    return reduce(lambda f, g: lambda x: f(g(x)), functions, lambda x: x)
```
Comment (lebaudantoine) [inline src/api/core/warren/utils.py:10]: Could we try calling`functions.reverse()` before 'head' to avoid the reverse order? I haven't tested it; it's up to you.

File: src/api/plugins/video/warren_video/api.py [hunk 7]
```diff
@@ -28,7 +28,7 @@ async def views(
         "client": lrs_client,
         "video_id": video_id,
         "date_range": DatetimeRange.parse_obj(filters),
-        "remove_duplicate_actors": unique,
+        "unique": unique,
```
Comment (lebaudantoine) [inline src/api/plugins/video/warren_video/api.py:31]: Nit: old debate, IMO unique is not explicit enough, as by definition each statement is already unique. But I agree, it feels weird to have a `unique` parameter that set a `remove_duplicate_actores` value.

**Review by lebaudantoine — COMMENTED** [review]
🤖🤗

File: src/api/core/warren/utils.py (hunk 6 above)
Comment (jmaupetit) [reply src/api/core/warren/utils.py:10]: The idea is to read them as you would declare them in a composition function series, _e.g._  `f(g(x))`.

File: src/api/core/warren/indicators/base.py (hunk 3 above)
Comment (jmaupetit) [reply src/api/core/warren/indicators/base.py:67]: Yup. That's on purpose (see my other comment). I am surprised that it's not integrated in Python stdlib. :man_shrugging:

File: src/api/core/warren/indicators/base.py (hunk 1 above)
Comment (jmaupetit) [reply src/api/core/warren/indicators/base.py:32]: Good point! This is silly, we should work on a copy.

File: src/api/core/warren/utils.py (hunk 6 above)
Comment (lebaudantoine) [reply src/api/core/warren/utils.py:11]: Why did you choose this approach over a sequential order? I'm more accustomed to a sequential, step-by-step, or even a 'pipeline' approach when dealing with consecutive data modifications. I draw inspiration from this [scikit-learn class](https://scikit-learn.org/stable/modules/generated/sklearn.pipeline.Pipeline.html).

File: src/api/plugins/video/warren_video/api.py (hunk 7 above)
Comment (lebaudantoine) [reply src/api/plugins/video/warren_video/api.py:31]: Alright your call !

File: src/api/core/warren/utils.py (hunk 6 above)
Comment (jmaupetit) [reply src/api/core/warren/utils.py]: To improve code readability, I've defined a `pipe` function instead that reverse functions application order. I think it's more explicit now.

File: src/api/plugins/video/warren_video/api.py [hunk 8]
```diff
@@ -25,10 +24,9 @@ async def views(
 ) -> DailyCounts:
     """Number of views for `video_id` in the `since` -> `until` date range."""
     indicator_kwargs = {
-        "client": lrs_client,
```
Comment (lebaudantoine) [inline src/api/plugins/video/warren_video/api.py:28]: nice I like it

File: src/api/core/warren/xapi.py [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+            ).hexdigest()
+
+        statements["actor.uid"] = statements.apply(get_uid, axis=1)
+        return statements
+
+    @staticmethod
+    def preprocess(statements: List[XAPI_STATEMENT]) -> pd.DataFrame:
+        """Denormalize raw statements, and add utility columns."""
+        return pipe(
+            StatementsTransformer.normalize_statements,
```
Comment (lebaudantoine) [inline src/api/core/warren/xapi.py:69]: Nit: What do you mean by `Denormalize` ? 
Should it be:
```suggestion
        """Normalize raw statements, and add utility columns."""
        return pipe(
            StatementsTransformer.normalize_statements,
```
To be consistent with method's naming.

File: src/api/core/warren/tests/test_xapi.py [hunk 10]
```diff
@@ -0,0 +1,198 @@
...
+    raw_statements = [
+        BaseXapiStatementFactory.build(
+            mutations=[{"timestamp": "2023-01-01T00:10:00.000000+00:00"}]
+        ).dict(),
+        BaseXapiStatementFactory.build(
+            mutations=[{"timestamp": "2023-01-03T00:10:00.000000+00:00"}]
+        ).dict(),
+    ]
+
+    statements = StatementsTransformer.normalize_statements(raw_statements)
```
Comment (lebaudantoine) [inline src/api/core/warren/tests/test_xapi.py:36]: Nit: Reading this method's call, is there any of the following options do you think is better:

- `StatementsTransformer.normalize(raw_statements)`
- `StatementsTransformer.normalize_raw(raw_statements)`

**Review by lebaudantoine — APPROVED** [review]
LGTM, thanks for this refactoring 👏

File: src/api/core/warren/xapi.py (hunk 9 above)
Comment (jmaupetit) [reply src/api/core/warren/xapi.py:69]: TBH I didn't noticed the docstring. :sweat_smile:

File: src/api/core/warren/tests/test_xapi.py (hunk 10 above)
Comment (jmaupetit) [reply src/api/core/warren/tests/test_xapi.py:36]: Good catch! `normalize()` would be a good candidate!

File: src/api/core/warren/indicators/base.py [hunk 11]
```diff
@@ -77,17 +17,32 @@ class BaseIndicator(ABC):
...
+        try:
+            return [
+                value
+                async for value in self.lrs_client.read(
+                    target=self.lrs_client.statements_endpoint,
+                    query=lrs_query or self.get_lrs_query(),
+                )
+            ]
+        except BackendException as exception:
+            raise LrsClientException("Failed to fetch statements") from exception
```
Comment (jmaupetit) [inline src/api/core/warren/indicators/base.py:45]: We should add tests for this.

**Review by lebaudantoine — APPROVED** [review]
